WordPress/src/main/java/org/wordpress/android/ui/jetpack/scan/ScanStateListItemBuilder.kt

```diff
@@ -0,0 +1,107 @@
+import org.wordpress.android.ui.utils.UiString.UiStringText
+import org.wordpress.android.viewmodel.ResourceProvider
+import java.util.concurrent.TimeUnit
+import javax.inject.Inject
+
+@Reusable
+class ScanStateListItemBuilder @Inject constructor(
+    private val htmlMessageUtils: HtmlMessageUtils,
+    private val resourceProvider: ResourceProvider
+) {
+    fun mapToScanState(
+        model: ScanStateModel,
+        site: SiteModel,
+        onScanButtonClicked: () -> Unit,
+        onFixAllButtonClicked: () -> Unit
+    ): ScanState {
+        return when (model.state) {
+            ScanStateModel.State.IDLE -> {
+                model.threats?.let { threats ->
+                    mapToThreatsFound(
+                        threats,
+                        site,
+                        onScanButtonClicked,
+                        onFixAllButtonClicked
+                    )
+                } ?: mapToThreatsNotFound(model, onScanButtonClicked)
+            }
+            ScanStateModel.State.SCANNING -> ScanScanningState()
+            ScanStateModel.State.PROVISIONING, ScanStateModel.State.UNAVAILABLE, ScanStateModel.State.UNKNOWN ->
+                ScanScanningState() // TODO: ashiagr filter out
+        }
+    }
+
+    private fun mapToThreatsFound(
+        threats: List<ThreatModel>,
+        site: SiteModel,
+        onScanButtonClicked: () -> Unit,
+        onFixAllButtonClicked: () -> Unit
+    ): ScanIdleState.ThreatsFound {
+        val fixableThreatsFound = threats.any { it.baseThreatModel.fixable != null }
+        return ScanIdleState.ThreatsFound(
+            scanDescription = buildThreatsFoundDescription(site, threats.size),
+            scanAction = buildScanButtonAction(R.string.scan_again, onScanButtonClicked),
+            fixAllAction = buildFixAllButtonAction(onFixAllButtonClicked).takeIf { fixableThreatsFound }
+        )
+    }
+
+    private fun mapToThreatsNotFound(scanStateModel: ScanStateModel, onScanButtonClicked: () -> Unit) =
+        ScanIdleState.ThreatsNotFound(
+            scanDescription = scanStateModel.mostRecentStatus?.startDate?.time?.let { buildLastScanDescription(it) }
+                ?: UiStringRes(R.string.scan_idle_manual_scan_description),
+            scanAction = buildScanButtonAction(R.string.scan_now, onScanButtonClicked)
+        )
+
+    private fun buildScanButtonAction(@StringRes titleRes: Int, onScanButtonClicked: () -> Unit) = ButtonAction(
+        title = UiStringRes(titleRes),
+        visibility = true,
+        onClicked = onScanButtonClicked
+    )
+
+    private fun buildFixAllButtonAction(onFixAllButtonClicked: () -> Unit) = ButtonAction(
+        title = UiStringRes(R.string.threats_fix_all),
+        visibility = true,
+        onClicked = onFixAllButtonClicked
+    )
+
+    private fun buildLastScanDescription(timeInMs: Long): UiStringResWithParams {
+        val durationInMs = System.currentTimeMillis() - timeInMs
```

Comment:
🔍 Nitpick: We might want to use injectable DateProvider here to make our tests deterministic. We might also want to add tests for the "scan_in_hour, ..." placeholders. If you plan to add those in future PRs feel free to ignore this.

Reply:
In 1769684